src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/FrameOfT.cs

```diff
@@ -28,15 +28,22 @@ public Frame(IHttpApplication<TContext> application, FrameContext frameContext)
         /// The resulting Task from this loop is preserved in a field which is used when the server needs
         /// to drain and close all currently active connections.
         /// </summary>
-        public override async Task RequestProcessingAsync()
+        public override async Task ProcessRequestsAsync()
         {
+            var dataRecieved = false;
+
             try
             {
                 while (!_requestProcessingStopping)
                 {
                     TimeoutControl.SetTimeout(_keepAliveTicks, TimeoutAction.CloseConnection);
 
-                    InitializeHeaders();
+                    if (dataRecieved)
```

Comment:
@pakrym and I discussed this. The bool wasn't there before, but this seemed cleaner than the `while (true) ... break` I had earlier.

Is there some alternative you're proposing. Do you think we should go back to delaying the call to RequestProcessingAsync() and re-adding the frame started TCS?

One benefit of this design is that dictionary initilization is elided for empty connections and delayed until a request is actually sent.